pkg/template/controller/templateinstance_finalizer.go

```diff
@@ -0,0 +1,259 @@
+		if v == templateapi.TemplateInstanceFinalizer {
+			needsFinalizing = true
+			break
+		}
+	}
+	if !needsFinalizing {
+		return nil
+	}
+
+	glog.V(4).Infof("TemplateInstanceFinalizer controller: syncing %s", key)
+
+	errs := []error{}
+	foreground := metav1.DeletePropagationForeground
+	deleteOpts := &metav1.DeleteOptions{PropagationPolicy: &foreground}
+	for _, o := range templateInstance.Status.Objects {
+		glog.V(5).Infof("attempting to delete object: %#v", o)
+
+		gv, err := schema.ParseGroupVersion(o.Ref.APIVersion)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("error parsing group version %s for object %#v: %v", o.Ref.APIVersion, o, err))
+			continue
+		}
+		gk := schema.GroupKind{
+			Group: gv.Group,
+			Kind:  o.Ref.Kind,
+		}
+
+		// if a resource type is removed, the template instance finalizer will
+		// never be able to clean up the template instance since it won't be
+		// able to map+delete all child resources that were previously created.
+		mapping, err := c.dynamicRestMapper.RESTMapping(gk, gv.Version)
+		if err != nil || mapping == nil {
+			errs = append(errs, fmt.Errorf("error mapping object %#v: %v", o, err))
+			continue
+		}
+
+		namespaced := mapping.Scope.Name() == meta.RESTScopeNameNamespace
+		namespace := ""
+		if namespaced {
+			namespace = o.Ref.Namespace
+		}
+
+		gvr := schema.GroupVersionResource{
+			Group:    gv.Group,
+			Version:  gv.Version,
+			Resource: mapping.Resource,
+		}
+
+		var resourceClient dynamic.DynamicResourceInterface
+		if namespaced {
+			resourceClient = c.client.NamespacedResource(gvr, namespace)
+		} else {
+			resourceClient = c.client.ClusterResource(gvr)
+		}
+		err = resourceClient.Delete(o.Ref.Name, deleteOpts)
+		if err != nil && !errors.IsNotFound(err) {
+			errs = append(errs, fmt.Errorf("error deleting object %#v with mapping %#v: %v", o, mapping, err))
+			continue
+		}
+	}
+	if len(errs) > 0 {
+		return kerrs.NewAggregate(errs)
+	}
+
+	templateInstanceCopy := templateInstance.DeepCopy()
+
+	newFinalizers := []string{}
+	for _, v := range templateInstanceCopy.Finalizers {
+		if v == templateapi.TemplateInstanceFinalizer {
+			continue
+		}
+		newFinalizers = append(newFinalizers, v)
+	}
+	templateInstanceCopy.Finalizers = newFinalizers
+
+	_, err = c.templateClient.Template().TemplateInstances(templateInstanceCopy.Namespace).UpdateStatus(templateInstanceCopy)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("TemplateInstanceFinalizer update failed: %v", err))
+		return err
+	}
+
+	return nil
+}
+
+// Run runs the controller until stopCh is closed, with as many workers as
+// specified.
+func (c *TemplateInstanceFinalizerController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	glog.V(2).Infof("TemplateInstanceFinalizer controller waiting for cache sync")
+	if !cache.WaitForCacheSync(stopCh, c.informerSynced) {
+		return
+	}
+
+	glog.V(2).Infof("Starting TemplateInstanceFinalizer controller")
```

Comment:
I recommend putting this at the info level so you know that you've sync'd.

Reply:
will do